extensions/Worker.Extensions.Storage.Queues/src/QueueMessageConverter.cs

```diff
@@ -0,0 +1,130 @@
+        {
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+        }
+
+        public ValueTask<ConversionResult> ConvertAsync(ConverterContext context)
+        {
+            var result = context?.Source switch
+            {
+                ModelBindingData binding => ConvertFromBindingData(context, binding),
+                _ => ConversionResult.Unhandled()
+            };
+
+            return new ValueTask<ConversionResult>(result);
+        }
+
+        private ConversionResult ConvertFromBindingData(ConverterContext context, ModelBindingData modelBindingData)
+        {
+            if (!IsQueueExtension(modelBindingData))
+            {
+                return ConversionResult.Unhandled();
+            }
+
+            try
+            {
+                QueueMessage queueMessage = ExtractQueueMessageFromBindingData(modelBindingData);
+                var result = ToTargetType(context.TargetType, queueMessage);
+
+                if (result is not null)
+                {
+                    return ConversionResult.Success(result);
+                }
+            }
+            catch (Exception ex)
+            {
+                return ConversionResult.Failed(ex);
+            }
+
+            return ConversionResult.Unhandled();
+
+        }
+
+        private bool IsQueueExtension(ModelBindingData bindingData)
+        {
+            if (bindingData?.Source is not Constants.QueueExtensionName)
+            {
+                _logger.LogTrace("Source '{source}' is not supported by {converter}", bindingData?.Source, nameof(QueueMessageConverter));
+                return false;
+            }
+
+            return true;
+        }
+
+        private object? ToTargetType(Type targetType, QueueMessage queueMessage) => targetType switch
+        {
+            Type _ when targetType == typeof(QueueMessage) => queueMessage,
+            Type _ when targetType == typeof(BinaryData) => ConvertMessageContentToBinaryData(queueMessage),
+            Type _ when targetType == typeof(JObject) => ConvertMessageContentToJObject(queueMessage),
+            _ => null
+        };
+
+        private QueueMessage ExtractQueueMessageFromBindingData(ModelBindingData modelBindingData)
+        {
+            if (modelBindingData.ContentType is not Constants.JsonContentType)
+            {
+                throw new NotSupportedException($"Unexpected content-type. Currently only '{Constants.JsonContentType}' is supported.");
+            }
+
+            try
+            {
+                JsonSerializerOptions options = new() { Converters = { new QueueMessageJsonConverter() } };
+                return modelBindingData.Content.ToObjectFromJson<QueueMessage>(options);
+            }
+            catch (JsonException ex)
+            {
+                // Easy to have the queue payload not deserialize properly. So give a useful error.
+                string msg = String.Format(CultureInfo.CurrentCulture,
```

Comment:
Using same message as the extension: https://github.com/Azure/azure-sdk-for-net/blob/056f7071a4b309a370c2ce7d52b2ef256f8dccf1/sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Queues/src/Triggers/UserTypeArgumentBindingProvider.cs#L74